### PR TITLE
   Add Grad-CAM visualization tool for DonkeyCar neural networks

### PR DESCRIPTION
**Pull request by rxhxm**
# Grad-CAM Visualization Tool for DonkeyCar

   ## Overview
   This PR adds a Gradient-weighted Class Activation Mapping (Grad-CAM) visualization tool to help better understand their neural network models. The tool provides a visual explanation of which parts of input images most influence the model's predictions, making it easier to debug and improve model performance.

   ## Features
   - Generate heatmap visualizations showing which regions of input images influence steering predictions
   - Process multiple frames to create visualization videos of model attention during driving
   - Exclude regions from visualization (such as dashboard/hood areas)
   - Support for various DonkeyCar model architectures
   - Command-line interface for easy integration with existing workflows

   ## Implementation
   The implementation consists of:
   - A core utility module with key Grad-CAM visualization functions
   - A command-line interface for batch processing
   - Example scripts and documentation

   ## Why This is Valuable
   Unlike the existing tools such as makemovie.py that simply show the model's predictions, Grad-CAM helps users understand *why* the model makes those predictions by visualizing which regions of the input image are most important to the model. This leads to:
   
   1. Better debugging of model behavior
   2. Improved model training through better understanding of network attention
   3. Detection of potential biases or issues in training data
   4. Enhanced user understanding of how neural networks process visual information

   ## Testing
   The tool has been tested on various DonkeyCar models with TensorFlow 2.15 and works with different model architectures including default DonkeyCar models.

   ## Documentation
   Documentation is included with examples in the README.md file within the module directory.

### CODE REVIEW

**Comment by mgagvani** [comment]
Just curious, how is this different than the `salient` flag in `makemovie` (https://github.com/autorope/donkeycar/blob/5483490aa3ef0fc8521447dd49a4de9f890830ec/donkeycar/parts/salient.py#L13) ?

File: donkeycar/contrib/gradcam/README.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# DonkeyCar Grad-CAM Visualization
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/README.md:1]: This documentation looks good. But it should go as a PR into the donkey_docs repository, so it will show up in the official documentation at docs.donkeycar.com

File: donkeycar/contrib/gradcam/__init__.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+"""
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/__init__.py:1]: If integrated into donkey, this file will become obsolete

File: donkeycar/contrib/gradcam/cli.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
+"""
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:1]: While the code in this file looks fine, the implementation should be integrated within the current `donkey makemovie` command and not within a separate script. We can extend the argument list for `makemovie` to add all the arguments you specify below.

File: donkeycar/contrib/gradcam/cli.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+            model = load_model(model_path, compile=False)
+            print("Model loaded successfully with compile=False")
+            return model
+        except Exception as e2:
+            print(f"Fatal error loading model: {e2}")
+            sys.exit(1)
+
+
+def find_images(data_path):
+    """
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:130]: This will not be needed when going through the `MakeMovie` class as all the data preparation is already in place.

File: donkeycar/contrib/gradcam/cli.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+        if image_paths:
+            print(f"Found {len(image_paths)} images with pattern {pattern}")
+            return image_paths
+
+    print("No images found with any pattern")
+    return []
+
+
+def load_metadata(data_path):
+    """
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:161]: Access to tub metadata and tub records is provided through the `Tub` class and hence this method will become obsolete.

File: donkeycar/contrib/gradcam/cli.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+        image_paths = sorted(image_paths, key=extract_number)
+        numbers = [extract_number(path) for path in image_paths]
+        print(f"Frame sequence: first={numbers[0]}, last={numbers[-1]}")
+        print(f"Sample sequence: {numbers[:10]} ...")
+    except Exception as e:
+        print(f"Could not sort images numerically: {e}")
+
+    # Process images
+    frames = []
+    for i, img_path in enumerate(image_paths):
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:254]: The code inside this loop is the core part of the PR. This is the bit that should be translated into `MakeMovie` such that we can choose between the current and your new version of displaying the salient features.  At that point all the tub data loading and file loading has already happened. This will greatly simplify your PR.

File: donkeycar/contrib/gradcam/cli.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+            prediction=prediction,
+            exclude_regions=exclude_regions
+        )
+
+        frames.append(vis_img)
+
+    return frames
+
+
+def create_video(frames, output_path, fps=20):
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:300]: This will become obsolete

File: donkeycar/contrib/gradcam/cli.py [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+        frame_bgr = cv2.cvtColor(frame, cv2.COLOR_RGB2BGR)
+        out.write(frame_bgr)
+
+    out.release()
+    print(f"Video saved to {output_path}")
+
+    return output_path
+
+
+def save_frames(frames, output_dir):
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:335]: As does this...

File: donkeycar/contrib/gradcam/cli.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+        # Save frame
+        frame_path = os.path.join(output_dir, f"frame_{i:04d}.jpg")
+        cv2.imwrite(frame_path, frame_bgr)
+
+    print(f"Individual frames saved to {output_dir}")
+
+    return output_dir
+
+
+def main():
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/cli.py:361]: And this...

File: donkeycar/contrib/gradcam/utils.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+            if len(coords) == 4:
+                regions.append(coords)
+    except Exception as e:
+        print(f"Error parsing exclude_regions: {e}")
+        return None
+
+    return regions
+
+
+def create_video(frames, output_path, fps=30):
```
Comment (DocGarbanzo) [inline donkeycar/contrib/gradcam/utils.py:225]: This will become obsolete.

**Review by DocGarbanzo — CHANGES_REQUESTED** [review]
@rxhxm - this looks like a good addition to our current salient method. However we cannot merge the code like this, simply because it has been written completely without using any of the donkey car framework, objects and methods. Any stand-alone script will very quickly stop working, because the code evolves over time, so delivering this new functionality within a stand-alone script is not the right thing to do.

However, it will be very easy to implement your new method within the `MakeMovie` class that currently implements the salient feature. This approach will integrate your new approach directly into the donkey car eco system, remove a lot of unnecessary code in your PR that deals with accessing data, and creating the movie from the individual frames, etc. 

It would be great if you could make that change.
